vulcand/registry.go

```diff
@@ -48,30 +49,33 @@ func NewRegistry(cfg Config, appName, ip string, port int) (*Registry, error) {
 		cfg.TTL = defaultRegistrationTTL
 	}
 
-	etcdCfg := cfg.Etcd
-	if etcdCfg == nil {
-		etcdCfg = &etcd.Config{Endpoints: []string{localEtcdProxy}}
+	if cfg.Etcd == nil {
+		cfg.Etcd = &etcd.Config{Endpoints: []string{localEtcdProxy}}
 	}
 
-	etcdClt, err := etcd.New(*etcdCfg)
+	client, err := etcd.New(*cfg.Etcd)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to create Etcd client, cfg=%v", etcdCfg)
+		return nil, errors.Wrapf(err, "failed to create Etcd client, cfg=%v", *cfg.Etcd)
 	}
 	ctx, cancelFunc := context.WithCancel(context.Background())
-	go func() {
-		for {
-			err := etcdClt.AutoSync(ctx, 10*time.Second)
-			if err == context.DeadlineExceeded || err == context.Canceled {
-				break
-			}
-			log.Errorf("Etcd auto sync failed: err=(%v)", err)
-		}
-	}()
-	etcdKeysAPI := etcd.NewKeysAPI(etcdClt)
+
+	// Grant a new lease for this client instance
+	resp, err := client.Grant(ctx, int64(cfg.TTL.Seconds()))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to grant a new lease, cfg=%v", *cfg.Etcd)
+	}
+
+	// Keep the lease alive for as long as we live
+	_, err = client.KeepAlive(ctx, resp.ID)
```

Comment:
If a client looses connection with `etcd` for longer then the lease TTL, the lease will expire in `etcd` and I suspect keep alive will not recreate it. So you need to account for that case. A nice way to test that scenario is using [toxyproxy](https://github.com/Shopify/toxiproxy) to simulate connection interruption with `etcd`.